### PR TITLE
[dynamo] do not raise an unimplemented error with boolean masking setitem

### LINKED ISSUE #134241: torch._dynamo.exc.Unsupported: boolean masking setitem backwards
### 🐛 Describe the bug

I don't know if this is a bug or a feature request, but when I use a boolean tensor to get and set values `torch._dynamo` throws an error. Minimal code to reproduce this:

```
import torch
import torch._dynamo

torch._dynamo.config.capture_dynamic_output_shape_ops = True


def test(y: torch.Tensor, z: torch.Tensor) -> torch.Tensor:
    x = torch.zeros_like(y, dtype=z.dtype)
    x[y] = z[y]
    x = x*2
    return x

z = torch.randn(10, 10, requires_grad=True).cuda()
y = (torch.randint(0, 1, (10, 10)) == 1).cuda()

test_comp = torch.compile(test, mode='reduce-overhead', fullgraph=True)
x_comp = test_comp(y, z)
loss_comp = x_comp.sum()
loss_comp.backward()
```

I am using `torch.__version__==2.5.0.dev20240821+cu118`

However, the interesting part is that I can avoid this error by first converting the boolean tensor `y` with `.nonzero()` into `int64` indices. This causes a cudagraph break due to 'incompatible_ops' when I compile with 'reduce-overhead', but doesn't throw any error with 'default' mode.

### Error logs

```
torch._dynamo.exc.Unsupported: boolean masking setitem backwards, see https://github.com/pytorch/pytorch/issues/114123

from user code:
   File "/home/richeek/GitHub/prolev/test.py", line 9, in test
    x[y] = z[y]

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```

### Minified repro

_No response_

### Versions

```
Collecting environment information...
PyTorch version: 2.5.0.dev20240821+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.11.9 (main, Apr 19 2024, 16:48:06) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-44-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 4090
GPU 1: NVIDIA GeForce RTX 4090

Nvidia driver version: 535.183.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             32
On-line CPU(s) list:                0-31
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 9 7950X 16-Core Processor
CPU family:                         25
Model:                              97
Thread(s) per core:                 2
Core(s) per socket:                 16
Socket(s):                          1
Stepping:                           2
CPU max MHz:                        5881.0000
CPU min MHz:                        400.0000
BogoMIPS:                           8983.19
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d
Virtualization:                     AMD-V
L1d cache:                          512 KiB (16 instances)
L1i cache:                          512 KiB (16 instances)
L2 cache:                           16 MiB (16 instances)
L3 cache:                           64 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-31
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Vulnerable: Safe RET, no microcode
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] numpy-quaternion==2023.0.3
[pip3] pytorch-triton==3.0.0+dedb7bdf33
[pip3] torch==2.5.0.dev20240821+cu118
[pip3] torch_cluster==1.6.3+pt23cu121
[pip3] torch_geometric==2.5.3
[pip3] torch_scatter==2.1.2+pt23cu118
[pip3] torch_sparse==0.6.18+pt23cu121
[pip3] torch_spline_conv==1.2.2+pt23cu121
[pip3] torchaudio==2.4.0.dev20240821+cu118
[pip3] torchvision==0.20.0.dev20240821+cu118
[pip3] triton==2.3.1
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] numpy-quaternion          2023.0.3                 pypi_0    pypi
[conda] pytorch-triton            3.0.0+dedb7bdf33          pypi_0    pypi
[conda] torch                     2.5.0.dev20240821+cu118          pypi_0    pypi
[conda] torch-cluster             1.6.3+pt23cu121          pypi_0    pypi
[conda] torch-geometric           2.5.3                    pypi_0    pypi
[conda] torch-scatter             2.1.2+pt23cu118          pypi_0    pypi
[conda] torch-sparse              0.6.18+pt23cu121          pypi_0    pypi
[conda] torch-spline-conv         1.2.2+pt23cu121          pypi_0    pypi
[conda] torchaudio                2.4.0.dev20240821+cu118          pypi_0    pypi
[conda] torchvision               0.20.0.dev20240821+cu118          pypi_0    pypi
[conda] triton                    2.3.1                    pypi_0    pypi
```

cc @ezyang @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by sudoRicheek**
Cudagraph breaks on boolean masking setitem, however the code runs fine. There is no need to raise an unimplemented error here, since it already warns that its an incompatible op.

Fixes #134241

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
Thanks for the contribution!
1) Just delete the code rather than commenting out
2) You need to sign the CLA
3) Is it possible to add a test for this?  Perhaps in `test/dynamo/test_repros.py`?

**Comment by sudoRicheek** [comment]
@jansel Thanks a lot! I'll do this

**Comment by jansel** [comment]
Looks like some tests are failing, any idea if that is a real error?

**Comment by sudoRicheek** [comment]
> Looks like some tests are failing, any idea if that is a real error?

I think its from this test: https://github.com/pytorch/pytorch/blob/96104db13259251a8bf0d24aba6c881eab2024c1/test/dynamo/test_export.py#L4518-L4571

I don't think it is needed anymore, since technically its not unsupported, it just causes a graph break

**Comment by jansel** [comment]
Want to update the test?

**Comment by henrylhtsang** [comment]
Hi, running into the unimplemented error as well. Would be nice if that gets fixed soon

**Comment by jansel** [comment]
Yeah I think we can remove that, though we should make sure the testcase does the correct thing with it removed.

**Comment by sudoRicheek** [comment]
Yes, I guess removing the assert should be fine. @jansel How do I check the test before making a commit? I haven't setup a torch dev environment.

**Comment by henrylhtsang** [comment]
> Yes, I guess removing the assert should be fine. @jansel How do I check the test before making a commit? I haven't setup a torch dev environment.

can you make the code changes and let the ci test it?

**Comment by henrylhtsang** [comment]
@sudoRicheek I need the changes. I would really appreciate if you can land it by end of week.

**Comment by sudoRicheek** [comment]
@henrylhtsang You can do the following changes to your local pytorch installation and it should work: https://github.com/pytorch/pytorch/pull/134902/commits/d60e37c7f3210ee62893309011824fcb716518c2

File: test/dynamo/test_export.py [hunk 1]
```diff
@@ -4565,11 +4565,6 @@ def forward(self, x, b, y):
     return pytree.tree_unflatten([x], self._out_spec)""",  # NOQA: B950
         )
 
-        with self.assertRaisesRegex(
-            torch._dynamo.exc.Unsupported, "boolean masking setitem backwards"
-        ):
-            gm, _ = torch._dynamo.export(fn)(x, b, y)
```
Comment (henrylhtsang) [inline test/dynamo/test_export.py:4571]: @sudoRicheek can we keep this line? Just remove the `with`

**Comment by henrylhtsang** [comment]
> @henrylhtsang You can do the following changes to your local pytorch installation and it should work: [d60e37c](https://github.com/pytorch/pytorch/commit/d60e37c7f3210ee62893309011824fcb716518c2)

Yeah, I mean I want it merged so it is in the trunk...

File: test/dynamo/test_export.py (hunk 1 above)
Comment (sudoRicheek) [reply test/dynamo/test_export.py:4571]: Just curious, if there is no assert, why do we need this line?

Comment (henrylhtsang) [reply test/dynamo/test_export.py:4571]: @sudoRicheek Good question. I guess it is good to show that it can run normally? Because if you remove both lines, the test is basically doing nothing.

**Review by henrylhtsang — APPROVED** [review]

**Comment by jansel** [comment]
(trying to check if those lint failures go away with a rebase)

**Review by jansel — APPROVED** [review]
